### ISSUE TITLE
i32 & 0xFFFF0000 -> literal out of range

### ISSUE THREAD

**Issue by Blackhole16** [rust-lang/rust#21979]
Having <a href=http://is.gd/G8Kn5z>this</a> code:
```rust
let i = 1 & 0xFFFF0000;
println!("{}" , i);
```
the following compiler warning is thrown: `warning: literal out of range for its type, #[warn(overflowing_literals)] on by default`

As this is not the case for u32 I think the compiler wants to tell me, that I should only use 31 bits. IMHO that is not very nice as I would need to use binary formatting, because Hex doesn't work :)

Wouldn't it be better not to have this warning? Or which other function does it serve?

Kind Regards,
BH16

**Comment by bluss** [comment]
If the compiler does not infer a particular integer type required by the code, it falls back to the default integer type `i32`. So the warning serves its purpose, it's telling you something is wrong! It's a sign that you need to explicitly say this:

```rust
let i: u32 = 1 & 0xFFFF0000;
```

Using u32 you will not see this warning — the literal fits the type.

**Comment by badboy** [comment]
If you tell the compiler explicitely to use an `u32`, it will be satisfied:

```
let i : u32 = 1 & 0xFFFF0000;
```

I can't find the exact rules written down atm, but if you don't tell the compiler about a type it will use some default, in this case `i32`

**Comment by steveklabnik** [comment]
It uses an `i32` if you don't annotate the literal. This doesn't seem to be a bug.
